snapshots/btrfs/btrfs.go

```diff
@@ -345,54 +304,45 @@ func (b *snapshotter) Remove(ctx context.Context, key string) (err error) {
 		readonly        bool
 	)
 
-	ctx, t, err := b.ms.TransactionContext(ctx, true)
-	if err != nil {
-		return err
-	}
-	defer func() {
-		if err != nil && t != nil {
-			if rerr := t.Rollback(); rerr != nil {
-				log.G(ctx).WithError(rerr).Warn("failed to rollback transaction")
-			}
+	err = b.ms.WithTransaction(ctx, true, func(ctx context.Context) error {
+		id, k, err := storage.Remove(ctx, key)
+		if err != nil {
+			return fmt.Errorf("failed to remove snapshot: %w", err)
+		}
+
+		switch k {
+		case snapshots.KindView:
+			source = filepath.Join(b.root, "view", id)
+			removed = filepath.Join(b.root, "view", "rm-"+id)
+			readonly = true
+		case snapshots.KindActive:
+			source = filepath.Join(b.root, "active", id)
+			removed = filepath.Join(b.root, "active", "rm-"+id)
+		case snapshots.KindCommitted:
+			source = filepath.Join(b.root, "snapshots", id)
+			removed = filepath.Join(b.root, "snapshots", "rm-"+id)
+			readonly = true
 		}
 
+		if err = btrfs.SubvolSnapshot(removed, source, readonly); err != nil {
+			removed = ""
+			return err
+		}
+
+		if err = btrfs.SubvolDelete(source); err != nil {
+			return fmt.Errorf("failed to remove snapshot %v: %w", source, err)
+		}
+		return nil
+	})
+
+	defer func() {
```

Comment:
I'd probably move this above the WithTransaction (and below the variable declarations) so you can see the cleanup case at the top of the method. Mostly a nit though